pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,550 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package photon
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/vmware/photon-controller-go-sdk/photon"
+	"gopkg.in/gcfg.v1"
+	"io"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	k8stypes "k8s.io/kubernetes/pkg/types"
+	"log"
+	"os/exec"
+	"strings"
+)
+
+const (
+	ProviderName = "photon"
+)
+
+// Global variable pointing to photon client
+var Esxclient *photon.Client
+var logger *log.Logger = nil
+
+// overrideIP indicates if the hostname is overriden by IP address, such as when
+// running multi-node kubernetes using docker. In this case the user should set
+// overrideIP = true in cloud config file. Default value is false.
+var overrideIP bool = false
+
+// Photon is an implementation of cloud provider Interface for Photon Controller.
+type PCCloud struct {
+	cfg *PCConfig
+	// InstanceID of the server where this PCCloud object is instantiated.
+	localInstanceID string
+	// local $HOSTNAME
+	localHostname string
+	// hostname from K8S, could be overridden
+	localK8sHostname string
+	// Photon Project ID
+	projID string
```

Comment:
Today the cloud provider assumes that it only manages disks in a single Photon Controller project, right? I think that's an important assumption to call out in a comment.

As an aside, I think this is totally fine for how we deploy Kubernetes today, since we deploy it on a per-project basis. It's possible that it won't be okay for future use cases. No need to handle them now... just thinking out loud.